File: templates/skills/workflow/scripts/context_budget.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
上下文预算计算工具。

从 shared-utils.md 中提取的纯数学计算逻辑。
ContextGovernor 的阈值比较在此实现，决策应用保留为伪代码。
"""

from __future__ import annotations

import json
import sys
from typing import Any, Dict, List, Optional


# =============================================================================
# Token Estimation
# =============================================================================


def estimate_tokens(*contents: Optional[str]) -> int:
    """估算文本内容的 token 数量（字符数 / 4）。

    >>> estimate_tokens("Hello world", "Test")
    4
    >>> estimate_tokens(None, "")
    0
    """
    total_chars = 0
    for content in contents:
        if content:
            total_chars += len(content)
    return round(total_chars / 4)


# =============================================================================
# Task Complexity
# =============================================================================


def detect_complexity(
    actions_count: int,
    file_count: int,
    is_quality_gate: bool,
    has_structured_steps: bool,
) -> str:
    """检测任务复杂度。

    >>> detect_complexity(1, 1, False, False)
    'simple'
    >>> detect_complexity(1, 3, False, True)
    'complex'
    """
    if is_quality_gate or has_structured_steps or file_count > 1:
        return "complex"
    if actions_count > 2:
        return "medium"
    return "simple"


# =============================================================================
# Dynamic Task Limits
# =============================================================================


def calculate_max_tasks(
    complexity: str,
    usage_percent: float,
) -> int:
    """根据复杂度和上下文使用率计算最大连续任务数。

    >>> calculate_max_tasks("simple", 30)
    8
    >>> calculate_max_tasks("medium", 75)
    2
    >>> calculate_max_tasks("complex", 85)
    1
    """
    base_limit = {"simple": 8, "medium": 5, "complex": 3}.get(complexity, 5)

    if usage_percent >= 80:
        return 1
    if usage_percent >= 70:
        return max(2, base_limit - 3)
    if usage_percent >= 50:
        return max(3, base_limit - 1)
    return base_limit


# =============================================================================
# Budget Threshold Evaluation (纯数学比较)
# =============================================================================


def evaluate_budget_thresholds(
    projected_usage_percent: float,
    warning_threshold: float = 60,
    danger_threshold: float = 80,
    hard_handoff_threshold: float = 90,
) -> Dict[str, Any]:
    """评估上下文预算阈值，返回纯数学比较结果。

    注意：此函数只返回阈值状态，不做决策（决策由 AI 的 ContextGovernor 执行）。

    >>> evaluate_budget_thresholds(55)
    {'level': 'safe', 'at_warning': False, 'at_danger': False, 'at_hard_handoff': False, 'projected_usage_percent': 55}
    >>> evaluate_budget_thresholds(85)
    {'level': 'danger', 'at_warning': True, 'at_danger': True, 'at_hard_handoff': False, 'projected_usage_percent': 85}
    """
    at_warning = projected_usage_percent >= warning_threshold
    at_danger = projected_usage_percent >= danger_threshold
    at_hard_handoff = projected_usage_percent >= hard_handoff_threshold

    if at_hard_handoff:
        level = "hard_handoff"
    elif at_danger:
        level = "danger"
    elif at_warning:
        level = "warning"
    else:
        level = "safe"

    return {
        "level": level,
        "at_warning": at_warning,
        "at_danger": at_danger,
        "at_hard_handoff": at_hard_handoff,
        "projected_usage_percent": projected_usage_percent,
    }


def project_next_turn_cost(
    current_tokens: int,
    execution_cost: int = 8000,
    verification_cost: int = 5000,
    review_cost: int = 0,
    safety_buffer: int = 4000,
) -> Dict[str, int]:
    """估算下一执行单元的 projected token 成本。

    >>> project_next_turn_cost(50000)
    {'current': 50000, 'execution': 8000, 'verification': 5000, 'review': 0, 'safety': 4000, 'projected_total': 67000}
    """
    projected_total = (
        current_tokens + execution_cost + verification_cost + review_cost + safety_buffer
    )
    return {
        "current": current_tokens,
        "execution": execution_cost,
        "verification": verification_cost,
        "review": review_cost,
        "safety": safety_buffer,
        "projected_total": projected_total,
    }


# =============================================================================
# Visual Bars
# =============================================================================


def generate_context_bar(
    usage_percent: float,
    warning_threshold: float = 60,
    danger_threshold: float = 80,
) -> str:
    """生成上下文使用率的可视化进度条。

    >>> generate_context_bar(45)
    '[🟩🟩🟩🟩🟩🟩🟩🟩🟩░░░░░░░░░░░] 45%'
    """
    filled = round(usage_percent / 5)
    bar = ""
    for i in range(20):
        if i < filled:
            if i >= danger_threshold / 5:
                bar += "🟥"
            elif i >= warning_threshold / 5:
                bar += "🟨"
            else:
                bar += "🟩"
        else:
            bar += "░"
    return f"[{bar}] {round(usage_percent)}%"


# =============================================================================
# CLI Entry
# =============================================================================


def main() -> int:
    import argparse

    parser = argparse.ArgumentParser(description="上下文预算计算工具")
    sub = parser.add_subparsers(dest="command")

    # estimate
    p_est = sub.add_parser("estimate", help="估算 token 数量")
    p_est.add_argument("files", nargs="+", help="文件路径列表")

    # complexity
    p_cx = sub.add_parser("complexity", help="检测任务复杂度")
    p_cx.add_argument("--actions", type=int, default=1, help="动作数")
    p_cx.add_argument("--files", type=int, default=1, help="文件数")
    p_cx.add_argument("--quality-gate", action="store_true")
    p_cx.add_argument("--structured-steps", action="store_true")

    # max-tasks
    p_mt = sub.add_parser("max-tasks", help="计算最大连续任务数")
    p_mt.add_argument("--complexity", required=True, choices=["simple", "medium", "complex"])
    p_mt.add_argument("--usage", type=float, required=True, help="上下文使用率 (%)")

    # budget
    p_budget = sub.add_parser("budget", help="评估预算阈值")
    p_budget.add_argument("--projected-usage", type=float, required=True, help="预计使用率 (%)")
    p_budget.add_argument("--warning", type=float, default=60)
    p_budget.add_argument("--danger", type=float, default=80)
    p_budget.add_argument("--hard-handoff", type=float, default=90)

    # context-bar
    p_bar = sub.add_parser("context-bar", help="生成上下文进度条")
    p_bar.add_argument("--usage", type=float, required=True, help="使用率 (%)")

    args = parser.parse_args()

    if args.command == "estimate":
        contents = []
        for fp in args.files:
            try:
                with open(fp, "r", encoding="utf-8") as f:
                    contents.append(f.read())
            except (OSError, IOError):
                contents.append(None)
        tokens = estimate_tokens(*contents)
        print(json.dumps({"estimated_tokens": tokens}))

    elif args.command == "complexity":
        result = detect_complexity(args.actions, args.files, args.quality_gate, args.structured_steps)
        print(json.dumps({"complexity": result}))

    elif args.command == "max-tasks":
        result = calculate_max_tasks(args.complexity, args.usage)
        print(json.dumps({"max_consecutive_tasks": result}))

    elif args.command == "budget":
        result = evaluate_budget_thresholds(
            args.projected_usage, args.warning, args.danger, args.hard_handoff
        )
        print(json.dumps(result))

    elif args.command == "context-bar":
        bar = generate_context_bar(args.usage)
        print(json.dumps({"bar": bar}, ensure_ascii=False))

    else:
        parser.print_help()
        return 1

    return 0


if __name__ == "__main__":
    sys.exit(main())

File: core/utils/workflow/context_budget.js
#!/usr/bin/env node
/**
 * @file 上下文预算管理 - 提供 token 估算、复杂度检测、预算阈值评估和上下文进度条生成
 */

const fs = require('fs')

/**
 * 根据字符数粗略估算 token 数量（约 4 字符 = 1 token）
 * @param {...string} contents - 一个或多个文本内容
 * @returns {number} 估算的 token 数
 */
function estimateTokens(...contents) {
  const totalChars = contents.reduce((sum, content) => sum + (content ? String(content).length : 0), 0)
  return Math.round(totalChars / 4)
}

/**
 * 根据动作数、文件数等信号判断任务复杂度
 * @param {number} actionsCount - 动作数量
 * @param {number} fileCount - 涉及文件数量
 * @param {boolean} isQualityGate - 是否为质量关卡
 * @param {boolean} hasStructuredSteps - 是否包含结构化步骤
 * @returns {'simple'|'medium'|'complex'} 复杂度等级
 */
function detectComplexity(actionsCount, fileCount, isQualityGate, hasStructuredSteps) {
  if (isQualityGate || hasStructuredSteps || fileCount > 1) return 'complex'
  if (actionsCount > 2) return 'medium'
  return 'simple'
}

/**
 * 根据复杂度和当前使用率计算单次可执行的最大任务数
 * @param {string} complexity - 复杂度等级（simple/medium/complex）
 * @param {number} usagePercent - 当前上下文使用百分比
 * @returns {number} 最大可执行任务数
 */
function calculateMaxTasks(complexity, usagePercent) {
  const baseLimit = { simple: 8, medium: 5, complex: 3 }[complexity] || 5
  if (usagePercent >= 80) return 1
  if (usagePercent >= 70) return Math.max(2, baseLimit - 3)
  if (usagePercent >= 50) return Math.max(3, baseLimit - 1)
  return baseLimit
}

/**
 * 评估预计使用率是否触达各级预算阈值
 * @param {number} projectedUsagePercent - 预计使用百分比
 * @param {number} warningThreshold - 警告阈值（默认 60）
 * @param {number} dangerThreshold - 危险阈值（默认 80）
 * @param {number} hardHandoffThreshold - 强制交接阈值（默认 90）
 * @returns {Object} 包含 level 和各阈值命中状态的评估结果
 */
function evaluateBudgetThresholds(projectedUsagePercent, warningThreshold = 60, dangerThreshold = 80, hardHandoffThreshold = 90) {
  const atWarning = projectedUsagePercent >= warningThreshold
  const atDanger = projectedUsagePercent >= dangerThreshold
  const atHardHandoff = projectedUsagePercent >= hardHandoffThreshold
  const level = atHardHandoff ? 'hard_handoff' : atDanger ? 'danger' : atWarning ? 'warning' : 'safe'
  return {
    level,
    at_warning: atWarning,
    at_danger: atDanger,
    at_hard_handoff: atHardHandoff,
    projected_usage_percent: projectedUsagePercent,
  }
}

/**
 * 预估下一轮操作的 token 消耗总量
 * @param {number} currentTokens - 当前已用 token 数
 * @param {number} executionCost - 执行消耗（默认 8000）
 * @param {number} verificationCost - 验证消耗（默认 5000）
 * @param {number} reviewCost - 审查消耗（默认 0）
 * @param {number} safetyBuffer - 安全缓冲（默认 4000）
 * @returns {Object} 各项消耗明细和预计总量
 */
function projectNextTurnCost(currentTokens, executionCost = 8000, verificationCost = 5000, reviewCost = 0, safetyBuffer = 4000) {
  const projectedTotal = currentTokens + executionCost + verificationCost + reviewCost + safetyBuffer
  return {
    current: currentTokens,
    execution: executionCost,
    verification: verificationCost,
    review: reviewCost,
    safety: safetyBuffer,
    projected_total: projectedTotal,
  }
}

/**
 * 生成可视化的上下文使用率进度条字符串
 * @param {number} usagePercent - 使用百分比
 * @param {number} warningThreshold - 警告阈值（默认 60）
 * @param {number} dangerThreshold - 危险阈值（默认 80）
 * @returns {string} 带颜色 emoji 的进度条字符串，如 [🟩🟩🟨🟥░░] 45%
 */
function generateContextBar(usagePercent, warningThreshold = 60, dangerThreshold = 80) {
  const filled = Math.round(usagePercent / 5)
  let bar = ''
  for (let i = 0; i < 20; i += 1) {
    if (i < filled) bar += i >= dangerThreshold / 5 ? '🟥' : i >= warningThreshold / 5 ? '🟨' : '🟩'
    else bar += '░'
  }
  return `[${bar}] ${Math.round(usagePercent)}%`
}

/**
 * 解析 CLI 参数，提取子命令和选项
 * @param {string[]} argv - 命令行参数数组
 * @returns {{command: string, args: string[], options: Object}} 解析结果
 */
function parseArgs(argv) {
  const args = [...argv]
  const command = args.shift()
  const options = {}
  while (args.length && args[0].startsWith('--')) {
    const flag = args.shift()
    if (flag === '--actions') options.actions = Number(args.shift())
    else if (flag === '--files') options.files = Number(args.shift())
    else if (flag === '--quality-gate') options.qualityGate = true
    else if (flag === '--structured-steps') options.structuredSteps = true
    else if (flag === '--complexity') options.complexity = args.shift()
    else if (flag === '--usage') options.usage = Number(args.shift())
    else if (flag === '--projected-usage') options.projectedUsage = Number(args.shift())
    else if (flag === '--warning') options.warning = Number(args.shift())
    else if (flag === '--danger') options.danger = Number(args.shift())
    else if (flag === '--hard-handoff') options.hardHandoff = Number(args.shift())
  }
  return { command, args, options }
}

function main() {
  const { command, args, options } = parseArgs(process.argv.slice(2))
  if (command === 'estimate') {
    const contents = args.map((file) => {
      try {
        return fs.readFileSync(file, 'utf8')
      } catch {
        return null
      }
    })
    process.stdout.write(`${JSON.stringify({ estimated_tokens: estimateTokens(...contents) })}\n`)
    return
  }
  if (command === 'complexity') {
    process.stdout.write(`${JSON.stringify({ complexity: detectComplexity(options.actions || 1, options.files || 1, Boolean(options.qualityGate), Boolean(options.structuredSteps)) })}\n`)
    return
  }
  if (command === 'max-tasks') {
    process.stdout.write(`${JSON.stringify({ max_consecutive_tasks: calculateMaxTasks(options.complexity, options.usage) })}\n`)
    return
  }
  if (command === 'budget') {
    process.stdout.write(`${JSON.stringify(evaluateBudgetThresholds(options.projectedUsage, options.warning, options.danger, options.hardHandoff))}\n`)
    return
  }
  if (command === 'context-bar') {
    process.stdout.write(`${JSON.stringify({ bar: generateContextBar(options.usage) })}\n`)
    return
  }
  process.stderr.write('Usage: node context_budget.js <estimate|complexity|max-tasks|budget|context-bar> ...\n')
  process.exitCode = 1
}

module.exports = {
  estimateTokens,
  detectComplexity,
  calculateMaxTasks,
  evaluateBudgetThresholds,
  projectNextTurnCost,
  generateContextBar,
}

if (require.main === module) main()
